### PR TITLE
add DeepPick type

### PR DESCRIPTION
**Pull request by cyberbiont**
this adds DeepPick type and tests for it, see
https://github.com/krzkaczor/ts-essentials/issues/168

DeepOmitModify helper class was  changed to FilterModify since now it is used by both DeepOmit and DeepPick types.
I also changed `Filter` type argument to `F` for clarity.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,1082 @@
+{
```
Comment (quezak) [inline package-lock.json:1]: pls remove this file, we use yarn here :)

Comment (cyberbiont) [reply package-lock.json]: Sorry, haven't noticed. Done!

**Comment by quezak** [comment]
@krzkaczor what do you think? IMO good for merge

File: test/index.ts [hunk 2]
```diff
@@ -375,6 +384,22 @@ function testDeepOmit2() {
...
+    id: string;
+    age: number;
+    name?: string;
+  };
+  type Omitted = {
+    id: string;
+    name?: string;
+  };
+
+  type Result = DeepOmit<OptionalProperty, { age: never }>;
```
Comment (krzkaczor) [inline test/index.ts:398]: shouldnt it be `DeepPick`?

Comment (cyberbiont) [reply test/index.ts:398]: Yes is should...

Update: upon consideration I discarded additional `testDeepOmit2` and `test deepPick` functions altogether and integrated all checks in one function `testDeepOmitAndDeepPick()` to make everything more neat. 
Test for commented lines will fail in the moment, this is what will be fixed by next PR: https://github.com/krzkaczor/ts-essentials/issues/176
Or should we do it all at once?

Comment (cyberbiont) [reply test/index.ts]: As for `true` instead of `never` , I thought along these lines too.
If we consider that Filter type may be used by `DeepPick` and `DeepOmit` simultaneously, `true` seems more adequate. But there's one implication that i found, is we rewrite the code using 'true', 'never' will also still work. So mixing both in Filter becomes possible. One also will be able to write filter types for DeepPick using `true` and filter types for DeepOmit using `never`, if that is his preference.

Comment (quezak) [reply test/index.ts]: Before we finally went with `never` I originally proposed the filter class to use `true`, so I'd have nothing against it, it still fits intuitively (you're selecting the properties to be ommited).
@krzkaczor?

Comment (krzkaczor) [reply test/index.ts]: So I would propose on making a breaking change and then `DeepOmit` should use `false` and the new type, `DeepPick` should use `true`. Does it make sense? Or should we simply use just `true` in both cases?

Comment (quezak) [reply test/index.ts]: `true` in both cases, the whole point of the breaking change is to use the same type here

**Comment by krzkaczor** [comment]
@cyberbiont so the last change would be to make both types use `true` instead of `never`. Would you like to work on this and finish it?

**Comment by owen2345** [comment]
Could u pls add this feature? It is very important in my project.
Thanks!

**Comment by cyberbiont** [comment]
@krzkaczor Sorry, almost forgot about this one. I have made the stipulated changes,  the tests are passing. I think it's ready to be merged.

**Comment by ali-habibzadeh** [comment]
What are the chances of merging this? What's the blocker?

**Comment by Beraliv** [comment]
@cyberbiont do you have a chance to finish it?

**Comment by cyberbiont** [comment]
@Beraliv, it was already finished, but obviously, instead of merging the PR some changes have been introduced that lead to merge conflicts.
OK I will see what i can do to fix it

**Comment by Beraliv** [comment]
> @Beraliv, it was already finished, but obviously, instead of merging the PR some changes have been introduced that lead to merge conflicts.
> OK I will see what i can do to fix it

If you can update it, that would be awesome

Please let me know if I can help you with it

**Comment by cyberbiont** [comment]
@Beraliv  Merge conflict fixed, you can take a look and merge the PR if it looks fine to you.

**Comment by Beraliv** [comment]
> @Beraliv  Merge conflict fixed, you can take a look and merge the PR if it looks fine to you.

I will have a look at it tomorrow evening 👌

Thank you again for the help! I really appreciate you spend time on it

**Comment by Beraliv** [comment]
@ali-habibzadeh @owen2345 I will release it in 8.1.0 tomorrow or the day after tomorrow

File: README.md [hunk 3]
```diff
@@ -27,54 +27,48 @@ If you use any [functions](https://github.com/krzkaczor/ts-essentials/blob/maste
 
 `ts-essentials` is a set of high-quality, useful TypeScript types that make writing type-safe code easier.
 
-- [Install](#Install)
```
Comment (Beraliv) [inline README.md:30]: can you please revert the changes here?

it looks a bit annoying as you don't really change anything here

if we want to run all the files in Prettier, we can do it in the different PR

WDYT?

File: README.md [hunk 4]
```diff
@@ -427,6 +421,42 @@ NOTE
 - `DeepOmit` works fine with `Array`s and `Set`s. When applied to a `Map`, the filter is only applied to its value.
 - If there exists any property in the filter which is not in the original type, an error will occur.
 
+### DeepPick
+
+Recursively pick deep properties according to key names.
+
+This type works as complementary type to DeepOmit, in the similar way like Exclude and Extract types complement each other.
```
Comment (Beraliv) [inline README.md:428]:
```suggestion
This type works as complementary type to `DeepOmit`, in the similar way like `Exclude` and `Extract` types complement each other.
```

File: README.md [hunk 5]
```diff
@@ -427,6 +421,42 @@ NOTE
...
+  gender: string;
+  students: { name: string; score: number }[];
+}
+```
+
+```typescript
+type TeacherSimple = DeepPick<
+  Teacher,
+  {
+    gender: never;
```
Comment (Beraliv) [inline README.md:446]: Now we have `true` filter instead of `never` AFAIK

```suggestion
    gender: true;
```

Don't forget to change it below too

File: lib/types.ts [hunk 6]
```diff
@@ -203,73 +203,130 @@ export type PickKeys<T, P> = Exclude<keyof PickProperties<T, P>, undefined>;
 /** Recursively omit deep properties */
 // explicitly mentioning optional properties, to work around TS making them required
 // see https://github.com/krzkaczor/ts-essentials/issues/118
-export type DeepOmit<T extends DeepOmitModify<Filter>, Filter> = T extends Builtin
+export type DeepOmit<T extends FilterModify<F>, F> = T extends Builtin
```
Comment (Beraliv) [inline lib/types.ts:206]: I would revert it to `Filter` as then you will have nicer message if you're doing something wrong or have an issue

```suggestion
export type DeepOmit<T extends FilterModify<Filter>, Filter> = T extends Builtin
```

File: lib/types.ts [hunk 7]
```diff
@@ -203,73 +203,130 @@ export type PickKeys<T, P> = Exclude<keyof PickProperties<T, P>, undefined>;
...
+            : T[K] extends FilterModify<F[K]>
+            ? DeepOmit<T[K], F[K]>
             : T[K];
         },
         never
       >;
-type DeepOmitModify<T> =
+
+/** Recursively pick deep properties */
+export declare type DeepPick<T extends FilterModify<F>, F> = T extends Builtin
```
Comment (Beraliv) [inline lib/types.ts:264]: Same here

```suggestion
export declare type DeepPick<T extends FilterModify<Filter>, Filter> = T extends Builtin
```

File: README.md [hunk 8]
```diff
@@ -427,6 +421,42 @@ NOTE
...
 
+### DeepPick
+
+Recursively pick deep properties according to key names.
+
+This type works as complementary type to DeepOmit, in the similar way like Exclude and Extract types complement each other.
+
+The filter syntax is the same as for the DeepPick, so one filter can be used to obtain both DeepPick and DeepOmit types from it.
+
+The properties to be picked completely should be defined as `never`. For the properties you want to partially pick, you should recursively define the sub-properties to be picked.
```
Comment (Beraliv) [inline README.md:432]: > The properties to be picked completely should be defined as `never`

That's not true, you can also have `true` as a picking indicator

**Comment by Beraliv** [comment]
@cyberbiont all in all, looks good for me, thank you for the help! 🔥

Also I found that we don't have a lot of test cases and it doesn't really work with all the cases, so I will merge it to the feature  branch `feat/deep-pick` and continue working on it there until it's ready regarding all the tests that I will add

The playground for checking the errors that I found – https://tsplay.dev/wgLaBN

**Comment by Beraliv** [comment]
@all-contributors please add @cyberbiont for code
